### PR TITLE
WIP update discovery to work with new multinic core

### PR DESCRIPTION
**Pull request by GregSutcliffe**
OK, here's my minimal changes to fix discovery for multinic support. Also needs theforeman/foreman#2116. Seems to work for both manual and auto-provisioning on latest core/develop.

I'm not sure how to make this compatible with 1.7 - most could be wrapped in version tests or .try() calls I guess?

I'll check back tomorrow to see how the tests are going :)

I also think we'll need to revisit this once the new multinic support has settled down and discuss how we want to consume it in Discovery.

### CODE REVIEW

File: app/models/host/discovered.rb [hunk 1]
```diff
@@ -9,8 +9,6 @@ class Host::Discovered < ::Host::Base
   belongs_to :hostgroup
   has_one    :discovery_attribute_set, :foreign_key => :host_id, :dependent => :destroy
 
-  validates :mac, :uniqueness => true, :mac_address => true, :presence => true
-  validates :ip, :format => {:with => Net::Validations::IP_REGEXP}, :uniqueness => true
```
Comment (GregSutcliffe) [inline app/models/host/discovered.rb:13]: Note this is validated by the Nic model now.

File: app/models/host/discovered.rb [hunk 2]
```diff
@@ -92,8 +90,14 @@ def import_facts facts
...
   end
 
+  def setup_clone
+    # Nic::Managed needs this method but Discovered hosts shouldn't
+    # be doing orchestration anyway...
+    self
+  end
+
   def attributes_to_import_from_facts
-    super + [:ip]
```
Comment (GregSutcliffe) [inline app/models/host/discovered.rb:87]: IP isn't a Parser method now, we look at all the interfaces

File: app/models/host/discovered.rb (hunk 1 above)
Comment (orrabin) [inline app/models/host/discovered.rb:13]: it's not great to run the validations twice but if that's the only problem maybe keep this in for now if you are trying to make this compatible to 1.7?

Comment (GregSutcliffe) [inline app/models/host/discovered.rb:13]: You can't keep it, there's no ip or mac column in develop anymore, you get errors on validation (and really weird errors at that...)

File: app/models/host/discovered.rb [hunk 3]
```diff
@@ -178,4 +182,10 @@ def compute_resource
     false
   end
 
+  def lookup_value_match
```
Comment (domcleal) [inline app/models/host/discovered.rb:185]: might this be an issue in core?  If Nic::Managed is used on Host::Base and upwards then perhaps lookup_value_match should move from Host::Managed to Host::Base.

Comment (ohadlevy) [inline app/models/host/discovered.rb:185]: what is calling this in the first place?

File: app/models/host/discovered.rb [hunk 4]
```diff
@@ -92,8 +90,14 @@ def import_facts facts
     super
   end
 
+  def setup_clone
```
Comment (ohadlevy) [inline app/models/host/discovered.rb:93]: what is calling this in the first place?

File: app/views/discovered_hosts/_discovered_hosts_list.html.erb [hunk 5]
```diff
@@ -35,7 +35,7 @@
       <% if SETTINGS[:organizations_enabled] -%>
         <td class="hidden-tablet hidden-xs"><%= host.organization.try(:name) %></td>
       <% end -%>
-      <td class="hidden-tablet hidden-xs"><%= host.subnet %></td>
+      <td class="hidden-tablet hidden-xs"><%= host.primary_interface.subnet %></td>
```
Comment (ohadlevy) [inline app/views/discovered_hosts/_discovered_hosts_list.html.erb:38]: any reason why not to delegate ip, subnet etc to primary interface on the model ?

File: app/models/host/discovered.rb (hunk 3 above)
Comment (domcleal) [inline app/models/host/discovered.rb:185]: <pre>
app/models/nic/managed.rb
111:      LookupValue.where(:match => "fqdn=#{fqdn_was}").update_all(:match => host.send(:lookup_value_match))
</pre>

File: app/models/host/discovered.rb (hunk 4 above)
Comment (GregSutcliffe) [inline app/models/host/discovered.rb:93]: nic orchestration calls setup clone on the parent host

File: app/models/host/discovered.rb (hunk 3 above)
Comment (GregSutcliffe) [inline app/models/host/discovered.rb:185]: @domcleal i can see value in a baseline method for the STI classes to override, yeah. I think it's harmless to leave this here for now, as we genuinely don't want to do matches on Discovered Hosts (for now anyway). I'll try and send something to core next week.

File: app/views/discovered_hosts/_discovered_hosts_list.html.erb (hunk 5 above)
Comment (GregSutcliffe) [inline app/views/discovered_hosts/_discovered_hosts_list.html.erb:38]: Potentially, but it's just minor refactoring. This is fine for now.

Comment (lzap) [inline app/views/discovered_hosts/_discovered_hosts_list.html.erb:38]: Yeah out of topic here :-)

Looks sane to me. Is this ready for testing now or do you plan any other updates?

**Comment by lzap** [comment]
I see errors when testing this against the foreman PR:

```
  1) Error:
Api::V2::DiscoveredHostsControllerTest#test_auto_provision_all_no_rule_success:
IPAddr::AddressFamilyError: address family must be specified
    /home/lzap/.rbenv/versions/2.0.0-p247/lib/ruby/2.0.0/ipaddr.rb:475:in `initialize'
    /home/lzap/work/foreman/app/models/subnet.rb:97:in `new'
    /home/lzap/work/foreman/app/models/subnet.rb:97:in `block in subnet_for'
    /home/lzap/work/foreman/app/models/subnet.rb:97:in `each'
    /home/lzap/work/foreman/app/models/subnet.rb:97:in `subnet_for'
    /home/lzap/work/foreman_discovery/app/models/host/discovered.rb:110:in `populate_fields_from_facts'
    /home/lzap/work/foreman/app/models/host/base.rb:108:in `import_facts'
    /home/lzap/work/foreman_discovery/app/models/host/discovered.rb:90:in `import_facts'
    /home/lzap/work/foreman_discovery/app/models/host/discovered.rb:82:in `import_host_and_facts'
    /home/lzap/work/foreman_discovery/test/functional/api/v2/discovered_hosts_controller_test.rb:75:in `test_auto_provision_all_no_rule_success'

  2) Error:
Api::V2::DiscoveredHostsControllerTest#test_auto_provision_all_success:
IPAddr::AddressFamilyError: address family must be specified
    /home/lzap/.rbenv/versions/2.0.0-p247/lib/ruby/2.0.0/ipaddr.rb:475:in `initialize'
    /home/lzap/work/foreman/app/models/subnet.rb:97:in `new'
    /home/lzap/work/foreman/app/models/subnet.rb:97:in `block in subnet_for'
    /home/lzap/work/foreman/app/models/subnet.rb:97:in `each'
    /home/lzap/work/foreman/app/models/subnet.rb:97:in `subnet_for'
    /home/lzap/work/foreman_discovery/app/models/host/discovered.rb:110:in `populate_fields_from_facts'
    /home/lzap/work/foreman/app/models/host/base.rb:108:in `import_facts'
    /home/lzap/work/foreman_discovery/app/models/host/discovered.rb:90:in `import_facts'
    /home/lzap/work/foreman_discovery/app/models/host/discovered.rb:82:in `import_host_and_facts'
    /home/lzap/work/foreman_discovery/test/functional/api/v2/discovered_hosts_controller_test.rb:66:in `test_auto_provision_all_success'

  3) Error:
Api::V2::DiscoveredHostsControllerTest#test_auto_provision_no_rule_success:
IPAddr::AddressFamilyError: address family must be specified
    /home/lzap/.rbenv/versions/2.0.0-p247/lib/ruby/2.0.0/ipaddr.rb:475:in `initialize'
    /home/lzap/work/foreman/app/models/subnet.rb:97:in `new'
    /home/lzap/work/foreman/app/models/subnet.rb:97:in `block in subnet_for'
    /home/lzap/work/foreman/app/models/subnet.rb:97:in `each'
    /home/lzap/work/foreman/app/models/subnet.rb:97:in `subnet_for'
    /home/lzap/work/foreman_discovery/app/models/host/discovered.rb:110:in `populate_fields_from_facts'
    /home/lzap/work/foreman/app/models/host/base.rb:108:in `import_facts'
    /home/lzap/work/foreman_discovery/app/models/host/discovered.rb:90:in `import_facts'
    /home/lzap/work/foreman_discovery/app/models/host/discovered.rb:82:in `import_host_and_facts'
    /home/lzap/work/foreman_discovery/test/functional/api/v2/discovered_hosts_controller_test.rb:58:in `test_auto_provision_no_rule_success'

  4) Error:
Api::V2::DiscoveredHostsControllerTest#test_auto_provision_success:
IPAddr::AddressFamilyError: address family must be specified
    /home/lzap/.rbenv/versions/2.0.0-p247/lib/ruby/2.0.0/ipaddr.rb:475:in `initialize'
    /home/lzap/work/foreman/app/models/subnet.rb:97:in `new'
    /home/lzap/work/foreman/app/models/subnet.rb:97:in `block in subnet_for'
    /home/lzap/work/foreman/app/models/subnet.rb:97:in `each'
    /home/lzap/work/foreman/app/models/subnet.rb:97:in `subnet_for'
    /home/lzap/work/foreman_discovery/app/models/host/discovered.rb:110:in `populate_fields_from_facts'
    /home/lzap/work/foreman/app/models/host/base.rb:108:in `import_facts'
    /home/lzap/work/foreman_discovery/app/models/host/discovered.rb:90:in `import_facts'
    /home/lzap/work/foreman_discovery/app/models/host/discovered.rb:82:in `import_host_and_facts'
    /home/lzap/work/foreman_discovery/test/functional/api/v2/discovered_hosts_controller_test.rb:49:in `test_auto_provision_success'
```

**Comment by GregSutcliffe** [comment]
I've got this targeted to the wrong branch - closing and re-opening.

**Pull request closed without merging**